### PR TITLE
evm circuit changes for dual code hash

### PR DESCRIPTION
**Pull request by z2trillion**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -39,4 +39,4 @@ debug-assertions = true
 overflow-checks = true
 rpath = false
 lto = "thin"
-incremental = false
+incremental = true
```
Comment (lispc) [inline Cargo.toml:42]: to speedup linking?

Comment (z2trillion) [reply Cargo.toml:42]: sorry, yes. I should not have included this in the PR though.

**Comment by lispc** [comment]
some ideas on first round review

(1) i am considering rename poseidon_code_hash to code_hash, so we need less [cfg[feature="scroll"] ...     , and be more friendly when merging upstream
(2) it seems some changes in this branch is compatible with scroll-stable(l1 evm), so for these changes that will not break, maybe it is better to first split them and merge them into scroll-stable first.   I guess we have to maintain 2 branch at least for a while(scroll-stable, compatible with l1 evm, and a dual code hash branch), so pushing the "common" stuff into scroll-stable will reduce the maintain burden
(2) scroll-stable changes a bit recently, so need to merge it.

**Comment by noel2004** [comment]
The `CodeHash` trait is induced for the 'dynamic codehash interface', which enable external caller (i.e. zkevm-scroll) to provide an codehash implement for building `codeDb`. We do not need this feature anymore if we has bring poseidon codehash into bus-mapping explicitly.

I advice to prune this feature completely and simply the codes around `CodeDb`

**Comment by z2trillion** [comment]
> some ideas on first round review
> 
> (1) i am considering rename poseidon_code_hash to code_hash, so we need less [cfg[feature="dual_code_hash"] ... , and be more friendly when merging upstream 
> (2) it seems some changes in this branch is compatible with scroll-stable(l1 evm), so for these changes that will not break, maybe it is better to first split them and merge them into scroll-stable first. I guess we have to maintain 2 branch at least for a while(scroll-stable, compatible with l1 evm, and a dual code hash branch), so pushing the "common" stuff into scroll-stable will reduce the maintain burden (2) scroll-stable changes a bit recently, so need to merge it.

This is the approach I went with in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1316. `code_hash` will always be whatever the code db uses as the key for the byte codes and in the scroll branch, we will need to add a `keccak_code_hash` field.

**Comment by noel2004** [comment]
> > some ideas on first round review
> > (1) i am considering rename poseidon_code_hash to code_hash, so we need less [cfg[feature="dual_code_hash"] ... , and be more friendly when merging upstream
> > (2) it seems some changes in this branch is compatible with scroll-stable(l1 evm), so for these changes that will not break, maybe it is better to first split them and merge them into scroll-stable first. I guess we have to maintain 2 branch at least for a while(scroll-stable, compatible with l1 evm, and a dual code hash branch), so pushing the "common" stuff into scroll-stable will reduce the maintain burden (2) scroll-stable changes a bit recently, so need to merge it.
> 
> This is the approach I went with in [privacy-scaling-explorations#1316](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1316). `code_hash` will always be whatever the code db uses as the key for the byte codes and in the scroll branch, we will need to add a `keccak_code_hash` field.

We have had an ['maintained' branch](https://github.com/scroll-tech/zkevm-circuits/tree/feat/codehash-maintained) which applied dual-codehash patch on `scroll-stable`

**Pull request closed without merging**